### PR TITLE
Ensure exactly single delivery of query completion event

### PR DESCRIPTION
**Pull request by losipiuk**
Thoughtful part of this PR by @findepi. I did the final edit :)

### CODE REVIEW

**Comment by findepi** [comment]
see failures

**Comment by losipiuk** [comment]
> see failures

Yeah - I spent couple hours on those today :/

File: presto-tests/src/test/java/io/prestosql/execution/TestEventListenerBasic.java [hunk 1]
```diff
@@ -259,7 +259,7 @@ public void testReferencedTablesAndRoutines()
         assertEquals(routine.getAuthorization(), "user");
     }
 
-    @Test
+    @Test(invocationCount = 100)
```
Comment (findepi) [inline presto-tests/src/test/java/io/prestosql/execution/TestEventListenerBasic.java:262]: remember to remove

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -239,7 +239,7 @@ jobs:
           - "presto-oracle"
           - "presto-kudu"
           - "presto-phoenix,presto-iceberg,presto-druid"
-    timeout-minutes: 60
+    timeout-minutes: 180
```
Comment (findepi) [inline .github/workflows/ci.yml:242]: reminder

File: presto-main/src/main/java/io/prestosql/dispatcher/LocalDispatchQuery.java [hunk 3]
```diff
@@ -56,22 +56,27 @@
...
+    private final QueryMonitor queryMonitor;
     private final ClusterSizeMonitor clusterSizeMonitor;
 
     private final Executor queryExecutor;
 
     private final Consumer<QueryExecution> querySubmitter;
-    private final SettableFuture<DispatchStatus> submitted = SettableFuture.create();
+    private final SettableFuture<?> submitted = SettableFuture.create();
+
+    AtomicBoolean notificationSentOrGuaranteed = new AtomicBoolean(false);
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/dispatcher/LocalDispatchQuery.java:67]: - private final

- remove `false`, as its the default ("code style" i guess)

File: presto-main/src/main/java/io/prestosql/dispatcher/LocalDispatchQuery.java [hunk 4]
```diff
@@ -86,16 +91,16 @@ public LocalDispatchQuery(
...
+                    queryMonitor.queryImmediateFailureEvent(getBasicQueryInfo(), getFullQueryInfo().getFailureInfo());
+                }
             }
-
-            if (state.ordinal() > QueryState.DISPATCHING.ordinal()) {
-                // we went past dispatching phase; warking submitted as successful
-                submitted.set(DISPATCHED);
+            if (state.isDone()) {
+                submitted.set(null);
+                queryExecutionFuture.cancel(true);
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/dispatcher/LocalDispatchQuery.java:103]: Previously `queryExecutionFuture.cancel(true);` was done only for FAILED state. Let's maybe retain that behavior, as it seems not related

File: presto-main/src/main/java/io/prestosql/dispatcher/LocalDispatchQueryFactory.java [hunk 5]
```diff
@@ -110,7 +109,7 @@ public DispatchQuery createDispatchQuery(
                 warningCollector,
                 StatementUtils.getQueryType(preparedQuery.getStatement().getClass()));
 
-        queryMonitor.queryCreatedEvent(stateMachine.getBasicQueryInfo(Optional.empty()));
+
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/dispatcher/LocalDispatchQueryFactory.java]: too many empty line

(this is fixed in the TMP commit)

File: presto-main/src/main/java/io/prestosql/execution/SqlQueryManager.java [hunk 6]
```diff
@@ -323,8 +315,8 @@ private void enforceScanLimits()
             Optional<DataSize> limitOpt = getQueryMaxScanPhysicalBytes(query.getSession());
             if (maxQueryScanPhysicalBytes.isPresent()) {
                 limitOpt = limitOpt
-                    .flatMap(sessionLimit -> maxQueryScanPhysicalBytes.map(serverLimit -> Ordering.natural().min(serverLimit, sessionLimit)))
-                    .or(() -> maxQueryScanPhysicalBytes);
+                        .flatMap(sessionLimit -> maxQueryScanPhysicalBytes.map(serverLimit -> Ordering.natural().min(serverLimit, sessionLimit)))
+                        .or(() -> maxQueryScanPhysicalBytes);
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/execution/SqlQueryManager.java:319]: separate commit, or revert

**Review by findepi — APPROVED** [review]

File: presto-main/src/main/java/io/prestosql/dispatcher/LocalDispatchQuery.java (hunk 4 above)
Comment (losipiuk) [reply presto-main/src/main/java/io/prestosql/dispatcher/LocalDispatchQuery.java:103]: Actually a bit before my changes it was done for all finished states. Not entiryly sure why I changed that. But given fact that we are rolling back the protocol I introduced it feels natural to rollback this change too:
image

**Comment by losipiuk** [comment]
https://github.com/prestosql/presto/pull/6219 is a followup ensuring early delivery of query-created event.
